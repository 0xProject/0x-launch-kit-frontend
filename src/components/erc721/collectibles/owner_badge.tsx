import React from 'react';
import styled from 'styled-components';

const Badge = styled.div`
    align-items: center;
    background: ${props => props.theme.componentsTheme.cardBackgroundColor};
    border-radius: 16px;
    box-shadow: 0px 1px 4px rgba(0, 0, 0, 0.04);
    display: flex;
    height: 31px;
    justify-content: center;
    min-width: 84px;
    padding: 8px 12px;
    position: absolute;
    left: 10px;
    top: 10px;
`;

const BadgeValue = styled.span`
    color: ${props => props.theme.componentsTheme.cardTitleOwnerColor};
    font-size: 14px;
    font-weight: 400;
    line-height: 14px;
`;

export const OwnerBadge = () => {
    return (
        <Badge>
            <BadgeValue>You Own</BadgeValue>
        </Badge>
    );
};
