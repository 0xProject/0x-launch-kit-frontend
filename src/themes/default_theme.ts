import { Theme, ThemeModalStyle, ThemeProperties } from './commons';

const modalThemeStyle: ThemeModalStyle = {
    content: {
        backgroundColor: '#fff',
        borderColor: '#dedede',
        bottom: 'auto',
        display: 'flex',
        flexDirection: 'column',
        flexGrow: '0',
        left: 'auto',
        maxHeight: '90%',
        minWidth: '350px',
        overflow: 'hidden',
        padding: '16px',
        position: 'relative',
        right: 'auto',
        top: 'auto',
    },
    overlay: {
        alignItems: 'center',
        backgroundColor: 'rgba(0, 0, 0, 0.4)',
        display: 'flex',
        justifyContent: 'center',
        zIndex: '12345',
    },
};

const lightThemeColors: ThemeProperties = {
    background: '#f4f3f4',
    backgroundERC721: '#F9FAFC',
    borderColor: '#dedede',
    boxShadow: '0 10px 10px rgba(0, 0, 0, 0.1)',
    buttonBuyBackgroundColor: '#3CB34F',
    buttonCollectibleSellBackgroundColor: '#00AE99',
    buttonConvertBackgroundColor: '#fff',
    buttonConvertBorderColor: '#dedede',
    buttonConvertTextColor: '#474747',
    buttonErrorBackgroundColor: '#FF6534',
    buttonPrimaryBackgroundColor: '#002979',
    buttonQuaternaryBackgroundColor: '#00AE99',
    buttonSecondaryBackgroundColor: '#474747',
    buttonSellBackgroundColor: '#FF6534',
    buttonTertiaryBackgroundColor: '#F6851B',
    buttonTextColor: '#fff',
    cardImageBackgroundColor: '#EBF0F5',
    cardBackgroundColor: '#fff',
    cardBorderColor: '#dedede',
    cardTitleColor: '#000',
    cardTitleOwnerColor: '#3CB34F',
    chartColor: '#00AE99',
    darkBlue: '#002979',
    darkGray: '#474747',
    darkerGray: '#666',
    dropdownBackgroundColor: '#fff',
    dropdownBorderColor: '#dedede',
    dropdownTextColor: '#000',
    errorButtonBackground: '#FF6534',
    errorCardBackground: '#FAF4EF',
    errorCardBorder: '#F39E4B',
    errorCardText: '#F68C24',
    ethBoxActiveColor: '#002979',
    ethBoxBorderColor: '#dedede',
    ethSetMinEthButtonBorderColor: '#000',
    ethSliderThumbBorderColor: 'rgba(0, 0, 0, 0.142)',
    ethSliderThumbColor: '#fff',
    gray: '#808080',
    green: '#3CB34F',
    iconLockedColor: '#000',
    iconUnlockedColor: '#C4C4C4',
    inactiveTabBackgroundColor: '#f9f9f9',
    lightGray: '#B9B9B9',
    logoERC20Color: '#0029FF',
    logoERC20TextColor: '#000',
    logoERC721Color: '#00AE99',
    logoERC721TextColor: '#000',
    marketsSearchFieldBackgroundColor: '#eaeaea',
    marketsSearchFieldBorderColor: '#dedede',
    marketsSearchFieldTextColor: '#333',
    modalSearchFieldBackgroundColor: '#fff',
    modalSearchFieldBorderColor: '#fff',
    modalSearchFieldPlaceholderColor: '#DEDEDE',
    modalSearchFieldTextColor: '#000',
    myWalletLinkColor: '#333333',
    notificationActive: '#F8F8F8',
    notificationIconColor: '#474747',
    notificationsBadgeColor: '#ff6534',
    numberDecimalsColor: '#dedede',
    red: '#FF6534',
    rowActive: '#FBFDFF',
    rowOrderActive: '#dedede',
    simplifiedTextBoxColor: '#F9F9F9',
    stepsProgressCheckMarkColor: '#fff',
    stepsProgressStartingDotColor: '#000',
    stepsProgressStepLineColor: 'rgba(0, 0, 0, 0.1)',
    stepsProgressStepLineProgressColor: '#000',
    stepsProgressStepTitleColor: '#e6e6e6',
    stepsProgressStepTitleColorActive: '#000',
    tableBorderColor: '#dedede',
    tdColor: '#000',
    textColorCommon: '#000',
    textDark: '#666',
    textInputBackgroundColor: '#F9F9F9',
    textInputBorderColor: '#dedede',
    textInputTextColor: '#000',
    textLight: '#999',
    textLighter: '#666',
    thColor: '#B9B9B9',
    tooltipBackgroundColor: '#222',
    tooltipTextColor: '#fff',
    topbarBackgroundColor: '#fff',
    topbarBorderColor: '#dedede',
    topbarSeparatorColor: '#dedede',
};

export class DefaultTheme implements Theme {
    public componentsTheme: ThemeProperties;
    public modalTheme: ThemeModalStyle;
    constructor() {
        this.componentsTheme = lightThemeColors;
        this.modalTheme = modalThemeStyle;
    }
}
