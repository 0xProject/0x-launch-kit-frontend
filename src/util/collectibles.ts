import { BigNumber } from '@0x/utils';

import { Collectible } from './types';

export const getCollectiblePrice = (collectible: Collectible): BigNumber | null => {
    const { order } = collectible;
    if (order === null) {
        return null;
    }
    return order.takerAssetAmount;

    // try {
    //     const dutchAcutionData = getDutchAuctionData(order.makerAssetData);
    //     const { beginAmount, beginTimeSeconds } = dutchAcutionData;
    //     const endAmount = order.takerAssetAmount;
    //     const startTimeSeconds = order.expirationTimeSeconds;
    //     // Use y = mx + b (linear function)
    //     const m = endAmount.minus(beginAmount).dividedBy(startTimeSeconds.minus(beginTimeSeconds));
    //     const b = beginAmount.minus(beginTimeSeconds.multipliedBy(m));
    //     return m.multipliedBy(todayInSeconds()).plus(b);
    // } catch (err) {
    //     return order.takerAssetAmount;
    // }
};
