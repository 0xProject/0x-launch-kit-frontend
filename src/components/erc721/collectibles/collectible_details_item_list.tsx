import React, { HTMLAttributes } from 'react';
import styled from 'styled-components';

import { themeDimensions } from '../../../themes/commons';
import { ChevronRight } from '../../common/icons/chevron_right_icon';

interface Props extends HTMLAttributes<HTMLDivElement> {
    color: string;
    image: string;
    name: string;
    onClick: (event: React.MouseEvent<HTMLElement>) => any;
}

const CollectibleAssetWrapper = styled.div`
    border-bottom: 1px solid ${props => props.theme.componentsTheme.cardBorderColor};
    cursor: pointer;
    display: flex;
    padding: 12px ${props => props.theme.modalTheme.content.padding};

    &:hover {
        background-color: rgba(0, 0, 0, 0.03);
    }

    &:last-child {
        border-bottom: none;
    }
`;

const ImageWrapper = styled.div<{ color: string; image: string }>`
    background-color: ${props => props.color || props.theme.componentsTheme.cardBackgroundColor};
    background-image: url('${props => props.image}');
    background-position: 50% 50%;
    background-size: contain;
    border-radius: ${themeDimensions.borderRadius};
    border: 1px solid ${props => props.theme.componentsTheme.cardBorderColor};
    height: 72px;
    margin-right: 15px;
    width: 72px;
`;

const TextContainer = styled.div`
    align-items: center;
    display: flex;
    flex-grow: 1;
    padding: 0 10px 0 0;
    overflow: hidden;
`;

const Title = styled.h3`
    color: ${props => props.theme.componentsTheme.cardTitleColor};
    font-size: 16px;
    font-weight: 600;
    line-height: 1.2;
    margin: 0 0 5px;
    overflow: hidden;
    padding: 0;
    text-overflow: ellipsis;
    white-space: nowrap;

    &:last-child {
        margin-bottom: 0;
    }
`;

const ChevronContainer = styled.div`
    align-items: center;
    display: flex;
    flex-shrink: 0;
`;

export const ListItem: React.FC<Props> = (props: Props) => {
    const { onClick, color, image, name } = props;
    return (
        <CollectibleAssetWrapper onClick={onClick}>
            <ImageWrapper color={color} image={image} />
            <TextContainer>
                <Title>{name}</Title>
            </TextContainer>
            <ChevronContainer>
                <ChevronRight />
            </ChevronContainer>
        </CollectibleAssetWrapper>
    );
};
