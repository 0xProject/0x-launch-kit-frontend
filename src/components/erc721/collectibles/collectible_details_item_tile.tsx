import { BigNumber } from '@0x/utils';
import React, { HTMLAttributes } from 'react';
import styled from 'styled-components';

import { themeDimensions, themeFeatures } from '../../../themes/commons';

import { PriceBadge } from './price_badge';

interface Props extends HTMLAttributes<HTMLDivElement> {
    color: string;
    image: string;
    name: string;
    onClick: (event: React.MouseEvent<HTMLElement>) => any;
    price: BigNumber | null;
}

const CollectibleAssetWrapper = styled.div`
    background: ${props => props.theme.componentsTheme.cardBackgroundColor};
    border-radius: ${themeDimensions.borderRadius};
    border: 1px solid ${props => props.theme.componentsTheme.cardBorderColor};
    box-sizing: border-box;
    cursor: pointer;
    position: relative;
    transition: box-shadow 0.15s linear;

    &:hover {
        box-shadow: ${themeFeatures.boxShadow};
    }
`;

const ImageWrapper = styled.div<{ color: string; image: string }>`
    background-color: ${props => props.color || props.theme.componentsTheme.cardBackgroundColor};
    background-image: url('${props => props.image}');
    background-position: 50% 50%;
    background-size: contain;
    border-top-left-radius: ${themeDimensions.borderRadius};
    border-top-right-radius: ${themeDimensions.borderRadius};
    height: 272px;
`;

const Title = styled.h2`
    color: ${props => props.theme.componentsTheme.cardTitleColor};
    font-size: 14px;
    font-weight: 500;
    line-height: 1.2;
    margin: 0;
    overflow: hidden;
    padding: 10px 12px;
    text-overflow: ellipsis;
    white-space: nowrap;
`;

export const TileItem: React.FC<Props> = (props: Props) => {
    const { onClick, color, image, price, name } = props;
    return (
        <CollectibleAssetWrapper onClick={onClick}>
            <ImageWrapper color={color} image={image}>
                <PriceBadge price={price} />
            </ImageWrapper>
            <Title>{name}</Title>
        </CollectibleAssetWrapper>
    );
};
