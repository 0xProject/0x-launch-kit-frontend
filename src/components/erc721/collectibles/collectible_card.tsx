import { BigNumber } from '0x.js';
import React from 'react';
import { connect } from 'react-redux';
import { Link } from 'react-router-dom';
import styled from 'styled-components';

import { ERC721_APP_BASE_PATH } from '../../../common/constants';
import { getEthAccount } from '../../../store/selectors';
import { themeDimensions, themeFeatures } from '../../../themes/commons';
import { Collectible, StoreState } from '../../../util/types';

import { OwnerBadge } from './owner_badge';
import { PriceBadge } from './price_badge';

const CollectibleCardWrapper = styled(Link)`
    background: ${props => props.theme.componentsTheme.cardBackgroundColor};
    border-radius: ${themeDimensions.borderRadius};
    border: 1px solid ${props => props.theme.componentsTheme.cardBorderColor};
    box-sizing: border-box;
    cursor: pointer;
    overflow: hidden;
    position: relative;
    transition: box-shadow 0.15s linear;
    text-decoration: none;

    &:hover {
        box-shadow: ${themeFeatures.boxShadow};
    }
`;

const ImageWrapper = styled.div<{ color: string; image: string }>`
    background-clip: padding-box;
    background-color: ${props => props.color || props.theme.componentsTheme.cardImageBackgroundColor};
    background-image: url('${props => props.image}');
    background-position: 50% 50%;
    background-size: contain;
    height: 272px;
`;

const Title = styled.h2`
    color: ${props => props.theme.componentsTheme.cardTitleColor};
    font-size: 14px;
    font-weight: 500;
    line-height: 1.2;
    margin: 0;
    overflow: hidden;
    padding: 10px 12px;
    text-overflow: ellipsis;
    white-space: nowrap;
`;

const defaultHandleClick = (e: any) => undefined;

interface OwnProps {
    collectible: Collectible;
    price: BigNumber | null;
    onClick?: (e: any) => void;
}

interface StateProps {
    ethAccount: string;
}

type Props = StateProps & OwnProps;

class CollectibleCard extends React.Component<Props> {
    public render = () => {
        const { collectible, price, ethAccount, onClick, ...restProps } = this.props;
        const { currentOwner, tokenId, color, image, name } = collectible;
        const isOwner = currentOwner.toLowerCase() === ethAccount.toLowerCase();
        const ownerBadge = isOwner ? <OwnerBadge /> : null;

        return (
            <CollectibleCardWrapper
                {...restProps}
                id={tokenId}
                onClick={onClick || defaultHandleClick}
                to={`${ERC721_APP_BASE_PATH}/collectible/${tokenId}`}
            >
                <ImageWrapper color={color} image={image}>
                    {ownerBadge}
                    <PriceBadge price={price} />
                </ImageWrapper>
                <Title>{name}</Title>
            </CollectibleCardWrapper>
        );
    };
}

const mapStateToProps = (state: StoreState): StateProps => {
    return {
        ethAccount: getEthAccount(state),
    };
};

const CollectibleCardContainer = connect(
    mapStateToProps,
    {},
)(CollectibleCard);

export { CollectibleCard, CollectibleCardContainer };
