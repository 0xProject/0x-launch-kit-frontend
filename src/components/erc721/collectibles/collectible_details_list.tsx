import React, { HTMLAttributes } from 'react';
import { connect } from 'react-redux';

import { selectCollectible } from '../../../store/collectibles/actions';
import { getCollectiblePrice } from '../../../util/collectibles';
import { Collectible } from '../../../util/types';

import { ListItem } from './collectible_details_item_list';
import { TileItem } from './collectible_details_item_tile';

interface OwnProps extends HTMLAttributes<HTMLDivElement> {
    collectible: Collectible;
    isListItem?: boolean;
    onClick: () => any;
}

interface DispatchProps {
    updateSelectedCollectible: (collectible: Collectible) => any;
}

type Props = DispatchProps & OwnProps;

export const CollectibleOnList: React.FC<Props> = (props: Props) => {
    const { collectible, onClick, isListItem } = props;
    const { color, image, name } = collectible;
    const price = getCollectiblePrice(collectible);

    const handleAssetClick: React.EventHandler<React.MouseEvent> = (event: React.MouseEvent<HTMLElement>) => {
        event.preventDefault();
        onClick();
        try {
            props.updateSelectedCollectible(collectible);
        } catch (err) {
            window.alert(`Could not sell the specified order`);
        }
    };

    return isListItem ? (
        <ListItem onClick={handleAssetClick} color={color} image={image} name={name} />
    ) : (
        <TileItem onClick={handleAssetClick} color={color} image={image} price={price} name={name} />
    );
};

const mapDispatchToProps = (dispatch: any): DispatchProps => {
    return {
        updateSelectedCollectible: (collectible: Collectible) => dispatch(selectCollectible(collectible)),
    };
};

export const CollectibleOnListContainer = connect(
    null,
    mapDispatchToProps,
)(CollectibleOnList);
