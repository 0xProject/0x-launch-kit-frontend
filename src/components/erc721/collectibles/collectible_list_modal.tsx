import React from 'react';
import Modal from 'react-modal';
import { connect } from 'react-redux';
import styled, { withTheme } from 'styled-components';

import { selectCollectible } from '../../../store/collectibles/actions';
import { getUserCollectiblesAvailableToSell } from '../../../store/selectors';
import { Theme } from '../../../themes/commons';
import { filterCollectibleByName } from '../../../util/filterable_collectibles';
import { Collectible, StoreState } from '../../../util/types';
import { EmptyContent } from '../../common/empty_content';
import { CloseModalButton } from '../../common/icons/close_modal_button';
import { InputSearch } from '../common/input_search';

import { CollectibleOnListContainer } from './collectible_details_list';

interface StateProps {
    userCollectibles: { [key: string]: Collectible };
}

interface DispatchProps {
    updateSelectedCollectible: (collectible: Collectible | null) => any;
}

interface OwnProps {
    theme: Theme;
    isOpen: boolean;
    onModalCloseRequest: () => any;
}

type Props = OwnProps & DispatchProps & StateProps;

interface State {
    filterText: string;
}

const modalWidth = '328px';
const modalContentHeight = '400px';

const ModalContent = styled.div`
    height: ${modalContentHeight};
    margin-left: -${props => props.theme.modalTheme.content.padding};
    margin-right: -${props => props.theme.modalTheme.content.padding};
    margin-top: -${props => props.theme.modalTheme.content.padding};
    overflow: auto;
    width: 360px;
`;

const ModalTitleWrapper = styled.div`
    border-bottom: 1px solid ${props => props.theme.componentsTheme.borderColor};
    margin: -3px -${props => props.theme.modalTheme.content.padding} 15px;
    padding: 0 ${props => props.theme.modalTheme.content.padding} 13px;
`;

const ModalTitleTop = styled.div`
    align-items: center;
    display: flex;
    justify-content: space-between;
    margin-bottom: 12px;
`;

const ModalTitle = styled.h2`
    color: ${props => props.theme.componentsTheme.cardTitleColor};
    font-size: 14px;
    font-weight: 500;
    line-height: 1.2;
    margin: 0;
    padding: 0 15px 0 0;
`;

const SearchStyled = styled(InputSearch)`
    background-color: ${props => props.theme.componentsTheme.marketsSearchFieldBackgroundColor};
    border-color: ${props => props.theme.componentsTheme.marketsSearchFieldBackgroundColor};
    color: ${props => props.theme.componentsTheme.marketsSearchFieldTextColor};
    max-width: 100%;
    width: ${modalWidth};

    ::placeholder {
        text-align: left;
    }
`;

const CloseModalButtonStyle = styled(CloseModalButton)`
    margin: 0;
`;

const initialState = {
    filterText: '',
};

class CollectibleListModalContainer extends React.PureComponent<Props, State> {
    public state = {
        ...initialState,
    };

    public render = () => {
        const { theme, isOpen, userCollectibles } = this.props;
        const collectibles = Object.keys(userCollectibles).map(key => userCollectibles[key]);
        const { filterText } = this.state;
        const filteredCollectibles = filterCollectibleByName(collectibles, filterText);
        return (
            <Modal isOpen={isOpen} style={theme.modalTheme} onRequestClose={this._closeModal}>
                <ModalTitleWrapper>
                    <ModalTitleTop>
                        <ModalTitle>Select an item to sell</ModalTitle>
                        <CloseModalButtonStyle onClick={this._closeModal} />
                    </ModalTitleTop>
                    <SearchStyled placeholder={'Search Wallet'} onChange={this._handleSearchInputChanged} />
                </ModalTitleWrapper>
                <ModalContent>
                    {filteredCollectibles.length > 0 ? (
                        filteredCollectibles.map((item, index) => (
                            <CollectibleOnListContainer
                                collectible={item}
                                isListItem={true}
                                key={index}
                                onClick={this._closeModal}
                            />
                        ))
                    ) : (
                        <EmptyContent text={'No results found'} />
                    )}
                </ModalContent>
            </Modal>
        );
    };

    private readonly _closeModal = () => {
        this.props.onModalCloseRequest();
        this.setState(initialState);
    };

    private readonly _handleSearchInputChanged = (event: React.ChangeEvent<HTMLInputElement>) => {
        this.setState({ filterText: event.target.value || '' });
    };
}

const mapStateToProps = (state: StoreState): StateProps => {
    return {
        userCollectibles: getUserCollectiblesAvailableToSell(state),
    };
};

const mapDispatchToProps = (dispatch: any): DispatchProps => {
    return {
        updateSelectedCollectible: (collectible: Collectible | null) => dispatch(selectCollectible(collectible)),
    };
};

const CollectibleListModal = withTheme(
    connect(
        mapStateToProps,
        mapDispatchToProps,
    )(CollectibleListModalContainer),
);

export { CollectibleListModal };
