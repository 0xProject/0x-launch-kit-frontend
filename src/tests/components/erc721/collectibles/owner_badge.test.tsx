/**
 * @jest-environment jsdom
 */

import React from 'react';
import renderer from 'react-test-renderer';
import { ThemeProvider } from 'styled-components';

import { OwnerBadge } from '../../../../components/erc721/collectibles/owner_badge';
import { DefaultTheme } from '../../../../themes/default_theme';

const theme = new DefaultTheme();

describe('OwnerBadge', () => {
    it('OwnerBadge to match snapshot', () => {
        // given
        const ownerBadge = (
            <ThemeProvider theme={theme}>
                <OwnerBadge />
            </ThemeProvider>
        );

        // when
        const tree = renderer.create(ownerBadge).toJSON();

        // then
        expect(tree).toMatchSnapshot();
    });
});
