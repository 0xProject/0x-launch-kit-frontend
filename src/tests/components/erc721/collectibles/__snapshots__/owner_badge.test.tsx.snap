// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`OwnerBadge OwnerBadge to match snapshot 1`] = `
<div
  className="sc-bdVaJa ksUms"
>
  <span
    className="sc-bwzfXH hyHzmj"
  >
    You Own
  </span>
</div>
`;
